lib/collection/src/operations/types.rs

```diff
@@ -1100,31 +1103,31 @@ impl CollectionError {
         Self::StrictMode { description }
     }
 
-    pub fn rate_limit_exceeded(description: impl Into<String>) -> Self {
-        Self::RateLimitExceeded {
-            description: description.into(),
-        }
-    }
-
     pub fn rate_limit_error(
         rate_limit_error: RateLimitError,
-        cost: Option<usize>,
+        cost: usize,
         write_limit_type: bool, // false = read rate limit; true = write rate limit.
     ) -> Self {
         let rate_limiter_type = if write_limit_type { "Write" } else { "Read" };
-        let description = match rate_limit_error {
-            RateLimitError::Message(msg) => {
-                format!("{rate_limiter_type} rate limit exceeded, {msg}",)
+        let (description, retry_after) = match rate_limit_error {
+            RateLimitError::AlwaysOverBudget(msg) => {
+                let description = format!("{rate_limiter_type} rate limit exceeded, {msg}",);
+                // no point in retrying this one
+                (description, None)
             }
-            RateLimitError::TokenLeft(token_available) => {
-                if cost.is_some() && cost.unwrap() > 1 {
-                    format!("{rate_limiter_type} rate limit exceeded: Operation requires {} tokens but only {token_available} were available. Retry later", cost.unwrap())
-                } else {
-                    format!("{rate_limiter_type} rate limit exceeded, retry later")
-                }
+            RateLimitError::Retry(retry_error) => {
+                let RetryError {
+                    tokens_available,
+                    retry_after,
+                } = retry_error;
+                let description = format!("{rate_limiter_type} rate limit exceeded: Operation requires {cost} tokens but only {tokens_available} were available. Retry later");
```

Comment:
Not sure about this suggestion:

```suggestion
                let description = format!("{rate_limiter_type} rate limit exceeded: Operation requires {cost} tokens but only {tokens_available} were available. Retry after {:?retry_after}");
```